src/core/Akka.Cluster.Tests/ClusterSpec.cs

```diff
@@ -306,6 +308,140 @@ public void A_cluster_must_be_allowed_to_join_and_leave_with_local_address()
             }
         }
 
+        [Fact]
+        public void A_cluster_must_be_able_to_JoinAsync()
+        {
+            var timeout = TimeSpan.FromSeconds(10);
+
+            try
+            {
+                _cluster.JoinAsync(_selfAddress).Wait(timeout).Should().BeTrue();
+                LeaderActions();
+                // Member should already be up
+                _cluster.Subscribe(TestActor, ClusterEvent.InitialStateAsEvents, new[] { typeof(ClusterEvent.IMemberEvent) });
+                ExpectMsg<ClusterEvent.MemberUp>();
+
+            }
+            finally
+            {
+                _cluster.Shutdown();
+            }
+        }
+
+        [Fact]
+        public void A_cluster_must_be_able_to_prematurelly_cancel_JoinAsync()
+        {
+            var timeout = TimeSpan.FromSeconds(10);
+
+            try
+            {
+                var cancel = new CancellationToken(true);
+                var task = _cluster.JoinAsync(_selfAddress, cancel);
+
+                Assert.Throws<AggregateException>(() => task.Wait(timeout))
+                    .Flatten()
+                    .InnerException.Should().BeOfType<TaskCanceledException>();
+
+                task.IsCanceled.Should().BeTrue();
+            }
+            finally
+            {
+                _cluster.Shutdown();
+            }
+        }
+
+        [Fact]
+        public void A_cluster_JoinAsync_must_fail_if_could_not_connect_to_cluster()
+        {
+            var timeout = TimeSpan.FromSeconds(10);
+
+            try
+            {
+                _cluster.Subscribe(TestActor, ClusterEvent.InitialStateAsEvents, new[] { typeof(ClusterEvent.IMemberEvent) });
+
+                var nonexisting = Address.Parse($"akka.tcp://{_selfAddress.System}@127.0.0.1:9999/");
+                Assert.ThrowsAsync<ClusterJoinFailedException>(async () =>
+                {
+                    await _cluster.JoinAsync(nonexisting);
+                    LeaderActions();
+
+                    ExpectMsg<ClusterEvent.MemberRemoved>();
+                }).Wait(timeout);
+
+            }
+            finally
+            {
+                _cluster.Shutdown();
+            }
+        }
+
+        [Fact]
+        public void A_cluster_must_be_able_to_join_async_to_seed_nodes()
```

Comment:
@Horusiath need a spec here that tests for "what happens if this method is called when we're already connected to the cluster?"